src/deriveNativeScriptHash.c

```diff
@@ -0,0 +1,485 @@
+#include "deriveNativeScriptHash.h"
+#include "state.h"
+#include "textUtils.h"
+#include "uiScreens.h"
+
+static ins_derive_native_script_hash_context_t* ctx = &(instructionState.deriveNativeScriptHashContext);
+
+// Helper functions
+
+#define TRACE_WITH_CTX(message, ...) TRACE(message "level = %u, remaining scripts = %u", ##__VA_ARGS__, ctx->level, ctx->complexScripts[ctx->level].remainingScripts)
+
+static inline bool areMoreScriptsExpected()
+{
+	// if the number of remaining scripts is not bigger than 0, then this request
+	// is invalid in the current context, as Ledger was not expecting another
+	// script to be parsed
+	return ctx->complexScripts[ctx->level].remainingScripts > 0;
+}
+
+static inline bool isComplexScriptFinished()
+{
+	return ctx->level > 0 && ctx->complexScripts[ctx->level].remainingScripts == 0;
+}
+
+static inline void complexScriptFinished()
+{
+	while (isComplexScriptFinished()) {
+		ASSERT(ctx->level > 0);
+		ctx->level--;
+
+		ASSERT(ctx->complexScripts[ctx->level].remainingScripts > 0);
+		ctx->complexScripts[ctx->level].remainingScripts--;
+
+		TRACE_WITH_CTX("complex script finished, ");
+	}
+}
+
+static inline void simpleScriptFinished()
+{
+	ASSERT(ctx->complexScripts[ctx->level].remainingScripts > 0);
+	ctx->complexScripts[ctx->level].remainingScripts--;
+
+	TRACE_WITH_CTX("simple script finished, ");
+
+	if (isComplexScriptFinished()) {
+		complexScriptFinished();
+	}
+}
+
+// UI
+
+const char ui_native_script_header[7][27] = {"Script - key path", "Script - key", "Script - ALL", "Script - ANY", "Script - N of K", "Script - invalid before", "Script - invalid hereafter"};
```

Comment:
Using a fixed length of 27 is not memory optimal. What is the proper way of minimizing the memory footprint here? We tried omitting the length 27, and a couple of variations with the PIC macro, but apparently we did not do it correctly because it caused Ledger to crash.